### DISCUSSION TITLE
Lit & Friends Community Call #3

### DISCUSSION THREAD

**Discussion by e111077** [lit/lit#3975] [Announcements]
Heya Lit friends,

After a hiatus the Lit & Friends Community Call is back! Our _third_ Lit & Friends Community Call on **Wednesday July 12** at  **10am PDT** on Google Meet.

Please join members of the Lit team to talk about all things Lit and web components.

## Agenda

The agenda has currently not been finalized, please discuss on this GH discussion to suggest items to put on this agenda.

Topics don't need to be limited to Lit, though they should be relevant. They may include web component standards, tooling, and more. We will prioritize existing open issues, pull requests, and discussions in the Lit repo, and address other topics time-permitting.

If you have questions about or issues with contributing to the Lit library, you should join in on our weekly Open Eng meeting which can be found [here](https://calendar.google.com/calendar/event?action=TEMPLATE&tmeid=MHBsY2sxcXVsOHFxMGk5czhscWloYjg4MWZfMjAyMzA2MjlUMTcwMDAwWiBjXzVmNzNiMzg5OTg2ODM0MWFhMTllYmQxZDhmY2QyYzQ5Yjg4NGNmYmNiZTdmOGU2NTUyMjdjNWE3OTA1YzA2OTJAZw&tmsrc=c_5f73b3899868341aa19ebd1d8fcd2c49b884cfbcbe7f8e655227c5a7905c0692%40group.calendar.google.com&scp=ALL).

### Lit Community Call # 3
[Calendar Link](https://calendar.google.com/calendar/event?action=TEMPLATE&tmeid=N2xvYXM0OWtlMDVzNXM1ZGdmMHNrczI0ODIgY181ZjczYjM4OTk4NjgzNDFhYTE5ZWJkMWQ4ZmNkMmM0OWI4ODRjZmJjYmU3ZjhlNjU1MjI3YzVhNzkwNWMwNjkyQGc&tmsrc=c_5f73b3899868341aa19ebd1d8fcd2c49b884cfbcbe7f8e655227c5a7905c0692%40group.calendar.google.com)
Wednesday, July 12 · 10-11am
Google Meet joining info
Video call link: https://meet.google.com/vvk-npbp-fri
Or dial: (US) +1 763-296-1736‬ PIN: ‪582 809 127‬#
More phone numbers: https://tel.meet/vvk-npbp-fri?pin=2263284382488&hs=1

### Code of Conduct
This call is covered by our [Code of Conduct](https://github.com/lit/lit/blob/main/CODE_OF_CONDUCT.md)

**Comment by patorpey** [comment]
Noting that the calendar link for the Community Call may be incorrect. It appears to be the same as the Open Eng call on 06 July, rather than the Community Call on 12 July.

**Reply by e111077** [reply]
Thanks! Updated. I guess sometimes you just need to check something 6 times over instead of 5 🙃

**Comment by justinfagnani** [comment]
For the agenda, I think as a standing item we should review RFCs. This would also be a good time to review plans for Lit 3.0 and standard decorators.

**Reply by AndrewJakubowicz** [reply]
I put up a rough draft of the lit-labs/compiler RFC so we can optionally review it or mention it.
Link: https://github.com/lit/rfcs/pull/21

**Reply by Kilsyth0203** [reply]
thank you so much

**Comment by bennypowers** [comment]
I'd like to discuss Analyzer integrations

**Comment by AndrewJakubowicz** [comment]
Thank you everyone for a great call!
If you’re interested in joining the next call, please keep eyes on GitHub discussions, [Discord announcements](https://lit.dev/discord), and [Twitter](https://twitter.com/buildWithLit)!
We also have a [public calendar](https://calendar.google.com/calendar/embed?src=c_5f73b3899868341aa19ebd1d8fcd2c49b884cfbcbe7f8e655227c5a7905c0692%40group.calendar.google.com&ctz=America%2FLos_Angeles), which you can add via the small plus icon on the bottom right of the screen (after navigating to the calendar link).

Here are the notes from todays meeting:


## Agenda topics

- What the Lit core team is working on update
- Analyzer integrations
- lit-labs/context: discussing the late upgrading middle provider case
- Feature request discussion: Strong IDE typescript integration
- Question: Any plan to add something similar to prop validation?
- Wrap up

# Notes

## What the Lit core team is working on update

Speaker: Justin

- Core team is working on Lit 3.0. Lit 3.0’s main breaking change is removing IE11 support and modernizing some of the JavaScript. This lets us ship smaller bundles & has minor performance increases.
- Due to fortunate timing, TypeScript is about to land standard decorator metadata. Because this requires TS 5.2, this could require a breaking TypeScript version change. The timeline lines up so we are working on standard decorators for 3.0 ([RFC](https://github.com/lit/rfcs/pull/20)).
- We will also be graduating labs packages: React, Context, and Task.
- And we’re spinning up two new labs packages:
    1. lit-labs/signals ([RFC](https://github.com/lit/rfcs/pull/19)): Preact signals integration package.
    2. lit-labs/compiler ([RFC](https://github.com/lit/rfcs/pull/21)): Provide an optional build step for pre-compiling lit templates – moving one time prep work for templates to build time. Improves first render performance.
- If you are using the labs packages scheduled for graduation, you may see changes coming in the coming weeks. Please give feedback!
- Regarding the standard decorator RFC, standard decorators and experimental decorators cannot be mixed and matched with the same tsconfig. Anyone interested in using both should reach out on Discord.
    - Within Lit monorepo to test experimental and standard decorators, we needed to use two tsconfigs.
    - The Lit team will not be removing experimental decorators for a long time (years?) because the standard decorator emit is larger than experimental decorators. We want to see browser support.
- Will Lit 3.0 be supported in parallel with Lit 2.0?
    - Yes. They are interoperable with each other.
- Is it supported to use Lit 2.0 and Lit 3.0 components in the same application?
    - You can mix different versions of Lit in the same app without any build step. This is the power of conforming to the web component interface.
- Can I use Lit 2 & 3 simultaneously with import maps?
    - Import maps can support both versions simultaneously in the same app.
- How was the Lit 2.0 -> Lit 3.0 migration within Google?
    - Within google there were 2 or 3 component breakages upgrading from Lit 2 -> Lit 3 amongst the thousands of components.
- If you have no DEV mode warnings, upgrading to Lit 3 is straightforward and requires no changes. See the [Lit 3.0 upgrade guide](https://lit.dev/docs/v3/releases/upgrade/).
- [Material Web Components](https://github.com/material-components/material-web) is nearing an initial 1.0 release!
- We are also improving server side rendering.


## Analyzer integrations

Topic: Benny Powers (@bennypowers)

- Benny has been running the Lit Analyzer and found a private API within the Analyzer called [gen-manifest](https://github.com/lit/lit/tree/main/packages/labs/gen-manifest). Is there any intention to open up and make these API’s public?
- [Kevin] Yes, the package is only marked private to prevent it being released too early. However if it is providing benefit, we are accepting contributions and happy to release the `gen-manifest` package.
- Some background on our Lit CLI & Analyzer:
    - The Analyzer is a low level library to statically analyze JS source that includes Lit and third party manifests. The Analyzer should be useful for a lot of tools, e.g. generating wrapper components, generating a manifest.
    - The Lit CLI is a modular CLI where the commands are sub packages that can be lazily installed via npm. 
- Thank you for trying out the Analyzer on your own project Benny, and providing feedback and contributions!

## lit-labs/context: discussing the late upgrading middle provider case


- This agenda topic was a discussion of  https://github.com/lit/lit/issues/2763
- The problematic upgrade order:
    - A grandparent (provider) upgrades, followed by a child (consumer), followed by a parent (provider).
    - The provided context is the same for the grandparent and parent, but the child will incorrectly end up with the value from the grandparent.
- This issue can also occur in light dom based on import statement order.
- How was this worked around by Benny?
    - All events are cached, and when someone else comes online then all events are re-fired.
- This issue also exists in the community protocol, and a fix needs to be proposed.
- The Lit team should design a solution and get this in before lit-labs/context graduation.
- Benny: our context solution, such as it is: https://github.com/RedHat-UX/red-hat-design-system/blob/153cba3c70786094c068a68e8cb9361db428058b/lib/context/color/controller.ts#L36

[Justin] Is there a usability issue with ContextRoot because it is only an imperative API?

- Do we want a custom element that does this and can be dropped into a page?
- Lit team unsure of the usage of ContextRoot and looking for feedback.

## Feature missing when migrating from React -> Lit. Strong IDE typescript integration


- Problem:
    - When changing a components API in a codebase, the IDE doesn’t provide any guardrails resulting in hampered devex.
    - Plugins have not worked or have false flags.
    - Is this something that could be more strongly integrated?
    - Concrete example: User of a component doesn’t provide a required attribute isn’t caught in IDE.
- Discussion:
    - By spec, every HTML element can be created without providing any attributes – and should not throw in that case. Imagine the img tag without `src`. It does nothing.
    - There is a convention that works for some apps – if an attribute is declared in TypeScript with a `!`, then it can be flagged by the VsCode Lit plugin.
    - Lit plugin: https://marketplace.visualstudio.com/items?itemName=runem.lit-plugin
    - Long term this is why we are working on our own analyzer implementation.
    - Members of the call agree that this should be provided as a standard package.
    - The vision: Analyzer works on first party source code, type checking and template checking. Then uses the manifest for third party components. This is a long term vision and requires work.

## Question: Plans in the future to add something like prop validation? (coming from Vue/React)

- Userland solution:
    - We added a custom decorator to add property validation and found it useful.
    - The decorator will emit a warning if the property is invalid to not be invasive and not shut down the app.
- Discussion:
    - This is interesting because it could be extended to become a noop when built into production. Custom userland decorators work great with Lit decorators, so using your own custom decorators is totally encouraged.
    - Regarding including this in core, an RFC is encouraged.
    - As a tangent, the type that a decorator takes today is not actually for types, but for deserializing to JavaScript property values.
        - It could be possible in dev mode to validate the type not just from attribute deserialization but also when setting properties.
    - As a side note – the intended behavior could be different based on context for an invalid prop:
        - A reusable standalone component will want to be forgiving.
        - An application component may want to fail fast if an invariant is not met.
    - Could this validation be transferred via the manifest?
        - Manifest documents the type of a given field or attribute and references a symbol. Maybe this is enough?

## Wrap up

- Thank you so much for coming and being amazing!
- We will schedule another call and post these notes!
- Please continue the discussion, we’d love to chat more!
- If you’re interested in the next call please keep eyes on GitHub discussions & [Discord announcements](https://lit.dev/discord), & [Twitter](https://twitter.com/buildWithLit)!
